### ISSUE TITLE
Aren't in a loop and we're trying to reassign breaks and continues

### ISSUE THREAD

**Issue by timdp** [babel/babel#1051]
I'm seeing this error with Babel 4.7.13 when processing a pretty long script (which is actually still ES5 at this point):

```
Error: [filename]: Aren't in a loop and we're trying to reassign breaks and continues, something is going wrong here.
    at BlockScoping.buildHas (node_modules\babel-core\lib\babel\transformation\transformers\es6\block-scoping.js:548:15)
    at BlockScoping.build (node_modules\babel-core\lib\babel\transformation\transformers\es6\block-scoping.js:519:12)
    at BlockScoping.wrapClosure (node_modules\babel-core\lib\babel\transformation\transformers\es6\block-scoping.js:400:10)
    at BlockScoping.run (node_modules\babel-core\lib\babel\transformation\transformers\es6\block-scoping.js:283:12)
    at TraversalPath.BlockStatement (node_modules\babel-core\lib\babel\transformation\transformers\es6\block-scoping.js:99:18)
    at TraversalPath.call (node_modules\babel-core\lib\babel\traversal\path.js:127:26)
    at TraversalPath.visit (node_modules\babel-core\lib\babel\traversal\path.js:142:10)
    at TraversalContext.visitNode (node_modules\babel-core\lib\babel\traversal\context.js:34:22)
    at TraversalContext.visit (node_modules\babel-core\lib\babel\traversal\context.js:42:19)
    at Function.traverse.node (node_modules\babel-core\lib\babel\traversal\index.js:44:17)
    at TraversalPath.visit (node_modules\babel-core\lib\babel\traversal\path.js:159:18)
    at TraversalContext.visitNode (node_modules\babel-core\lib\babel\traversal\context.js:34:22)
    at TraversalContext.visit (node_modules\babel-core\lib\babel\traversal\context.js:51:28)
```

I was only passing `blacklist: ['strict']` originally, but it also happens from the command line, without any options.

I'd be happy to provide a minimal script that triggers the error, but how would I go about debugging this? From the looks of it, it's walking the AST, so I should probably be able to find the tree node it's choking on and map that back to the source?

**Comment by timdp** [comment]
I've managed to narrow it down a bit. At some point in the script, there's essentially this gem:

```js
this.func1 = function() {
  if (cond1) {
    for (...) {
      if (cond2) {
        function func2() {}
        function func3() {}
        func4(function() {
          func2();
        });
        break;
      }
    }
  }
};
```

At this point, I feel I should mention that I didn't write it.

Anyway, it looks like the combination of `function func2` and `break` is causing the error. If I comment out the callback in the `func4` call, transpilation works. Moreover, if I replace `function func2()` with `var func2 = function()`, that also works.

**Comment by sebmck** [comment]
Looks like this was some legacy code left over as there's no reason for that error to be thrown. I've added a test and removed that assertion. Fixed as of 4.7.16, thanks!
